### PR TITLE
feat: File System Abstraction local/AWS TDE-407

### PR DESCRIPTION
**Pull request by paulfouquet**
## Description
This is a start to build a File System Abstraction in Topo-Imagery container. This change focus on having a `read` and `write` method which abstract the fact that the files are local or on AWS S3.

## Changes
- `fs` provides a `read` and a `write` methods which allows local or AWS S3 path.
- A bit of refactoring on the python project structure.

### CODE REVIEW

File: scripts/aws/aws_helper.py [hunk 1]
```diff
@@ -79,6 +79,14 @@ def get_bucket_name_from_path(path: str) -> str:
 
 
 def parse_path(path: str) -> Tuple[str, str]:
```
Comment (blacha) [inline scripts/aws/aws_helper.py:81]: A tuple doesnt say what is what, so using something that knows that the first part is `bucket` would be easier for people to understand

```python
ret = parse_path("s3://foo/bar")
ret.bucket # foo
ret.key # /bar
```

File: scripts/files/fs_local.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+def write(destination: str, source: bytes) -> None:
+    """Write the source to the local destination file.
+
+    Args:
+        destination (str): The local path to the file to write.
+        source (bytes): The source file in bytes.
+    """
+    with open(destination, "wb") as buffer:
```
Comment (blacha) [inline scripts/files/fs_local.py:8]: NIT: You are opening a `file` not a `buffer`.

File: scripts/files/fs_s3.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    Args:
+        destination (str): The AWS S3 path to the file to write.
+        source (bytes): The source file in bytes.
+        needs_credentials (bool, optional): Tells if credentials are needed. Defaults to False.
+    """
+    if source is None:
+        get_log().error("write_s3_source_none", path=destination, error="The 'source' is None.")
+        raise Exception("The 'source' is None.")
+    bucket_name, key = parse_path(destination)
+    key = key[1:]
```
Comment (blacha) [inline scripts/files/fs_s3.py:46]: this is a odd statement why is this needed?

File: scripts/files/fs_s3.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    s3 = boto3.resource("s3")
```
Comment (blacha) [inline scripts/files/fs_s3.py:47]: You could simplify this to always calling `get_auth_s3_session(bucket_name, "write")` which then only gives you a authed s3 object if you need it.

File: scripts/files/fs_s3.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+def get_auth_s3_session(bucket_name: str) -> S3ServiceResource:
+    """Return a boto3 S3 Resource with the AWS credentials for a bucket.
+
+    Args:
+        bucket_name (str): The name of the bucket.
+
+    Returns:
+        S3ServiceResource: The boto3 S3 Resource.
+    """
+    credentials = get_credentials(bucket_name)
```
Comment (blacha) [inline scripts/files/fs_s3.py:24]: you could try reading/writing the location before creating credentials as the default creds may have enough power to read or write.

File: scripts/files/fs_s3.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    session = boto3.Session(
```
Comment (blacha) [inline scripts/files/fs_s3.py:25]: I think it would be better to use a specific AssumeRoleCredential helper similar to how it is done https://github.com/linz/lds-cache-client-py/blob/master/lds_cache/lds_cache_source.py#L35

File: scripts/files/fs_s3.py (hunk 5 above)
Comment (paulfouquet) [reply scripts/files/fs_s3.py:24]: That's what it's trying to do in the `read` or `write` methods actually

File: scripts/files/fs_s3.py (hunk 3 above)
Comment (paulfouquet) [reply scripts/files/fs_s3.py]: To remove the leading `/`. This will be put in the `parse_path` method instead.

File: scripts/files/fs_s3.py (hunk 6 above)
Comment (paulfouquet) [reply scripts/files/fs_s3.py]: I leave the refreshable credential for another ticket (TDE-235)

File: scripts/files/fs_s3.py (hunk 4 above)
Comment (paulfouquet) [reply scripts/files/fs_s3.py:56]: I leave that for when/if `write` roles will be exposed.

File: scripts/aws/aws_helper.py [hunk 7]
```diff
@@ -13,6 +13,7 @@
     Bucket = object
 
 Credentials = namedtuple("Credentials", ["access_key", "secret_key", "token"])
+S3_path = namedtuple("S3_path", ["bucket", "key"])
```
Comment (blacha) [inline scripts/aws/aws_helper.py:16]: Why named tuples? do you have reasoning for using them vs other container classes?

Also mixing Snake and TitleCase is a little odd,  what is the prefered method for naming Objects Title?

File: scripts/converters/format_source.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    """Due to Argo constraints if using the basemaps cli list command
+    the source has a string that contains a list that needs to be split.
+    example: ["[\"s3://test/image_one.tiff\", \"s3://test/image_two.tiff\"]"]
+    """
+    if len(source) == 1 and source[0].startswith("["):
+        try:
+            source_json: List[str] = json.loads(source[0])
+            return source_json
+        except json.JSONDecodeError as e:
+            get_log().debug("Decoding Json Failed", source=source, msg=e)
```
Comment (blacha) [inline scripts/converters/format_source.py:18]: should this really swallow the error? If a invalid input is detected should it not throw?

File: scripts/create_polygons.py [hunk 9]
```diff
@@ -46,43 +47,37 @@ def main() -> None:
...
+                    get_log().debug(f"- data_px_count was zero in create_mask function for the tif {mask_file}")
+                else:
+                    destination_file_name = os.path.splitext(source_file_name)[0] + ".geojson"
+                    temp_file_path = os.path.join(tmp_dir, destination_file_name)
+                    polygonize_command = f'gdal_polygonize.py -q "{mask_file}" "{temp_file_path}" -f GeoJSON'
+                    os.system(polygonize_command)
+
+                output_files.append(temp_file_path)
+        except ClientError as ce:
+            get_log().error("create_polygon_skip_file", error=str(ce))
```
Comment (blacha) [inline scripts/create_polygons.py:79]: we should fail the command if something fails?

File: scripts/files/fs_s3.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        if needs_credentials:
+            s3 = _get_s3_resource(s3_path.bucket)
+
+        s3_object = s3.Object(s3_path.bucket, key)
+        file: bytes = s3_object.get()["Body"].read()
+    except botocore.exceptions.ClientError as ce:
+        if not needs_credentials and ce.response["Error"]["Code"] == "AccessDenied":
+            get_log().debug("read_s3_needs_credentials", path=path)
+            return read(path, True)
+        if ce.response["Error"]["Code"] == "NoSuchBucket":
```
Comment (blacha) [inline scripts/files/fs_s3.py:92]: you should use the `except boto...exceptions....NoSuchBucket` style for this

File: scripts/aws/aws_helper.py (hunk 7 above)
Comment (paulfouquet) [reply scripts/aws/aws_helper.py:16]: > Why named tuples? do you have reasoning for using them vs other container classes?

1. I was using a `tuple` before and did not want to create a `class` when the purpose was only to name those properties.
2. I found this interesting topic, https://stackoverflow.com/questions/51671699/data-classes-vs-typing-namedtuple-primary-use-cases

> Also mixing Snake and TitleCase is a little odd, what is the prefered method for naming Objects Title?

This is purely a mistake, this should be `s3_path`, good spotting.

File: scripts/converters/format_source.py (hunk 8 above)
Comment (paulfouquet) [reply scripts/converters/format_source.py:18]: The way this change appears is really confusing. I haven't write this code. I just moved the file.

File: scripts/create_polygons.py (hunk 9 above)
Comment (paulfouquet) [reply scripts/create_polygons.py:79]: This one is interesting. Let's say it fails on reading the file from its source which is a s3 bucket. If for some reasons, some other files are on another bucket that we have access too, should we stop them to be processed?

File: scripts/aws/aws_helper.py (hunk 7 above)
Comment (blacha) [reply scripts/aws/aws_helper.py:16]: Just looking at the articile you may want `NamedTuple` then

```python
from collections import namedtuple
from typing import NamedTuple

Foo = namedtuple("Foo", ["a", "b"])
Foo("a","b").a # Any
Foo("a", 1).b # Any

Bar = NamedTuple("Bar", [("a", str), ("b", int)])
Bar("a", 1).a # str
Bar("a", 1).b # int
```

File: scripts/files/fs_s3.py (hunk 10 above)
Comment (paulfouquet) [reply scripts/files/fs_s3.py:92]: Ha, I removed my comment somehow. When using boto3 `Resources`, exceptions are a bit different than for `Client`. 
https://boto3.amazonaws.com/v1/documentation/api/latest/guide/error-handling.html#catching-exceptions-when-using-a-resource-client
I am not able to find the `AccessDenied` exception. Should I keep the existing behavior for `AccessDenied` and catch the exception using `except` on each specific exceptions for the other errors?

File: scripts/aws/aws_helper.py (hunk 7 above)
Comment (paulfouquet) [reply scripts/aws/aws_helper.py:16]: I should specify the type and use TitleCase then, if it is the convention

File: scripts/create_polygons.py (hunk 9 above)
Comment (blacha) [reply scripts/create_polygons.py:79]: I would process them all and just fail the command at the end

Comment (paulfouquet) [reply scripts/create_polygons.py:79]: Meaning `continue` to the next file when one fails, keep the information that it failed and at the end fail the command if at least one failed?

File: scripts/converters/format_source.py (hunk 8 above)
Comment (paulfouquet) [reply scripts/converters/format_source.py:18]: I'll fix that in an other PR

File: scripts/create_polygons.py (hunk 9 above)
Comment (paulfouquet) [reply scripts/create_polygons.py:79]: This should not be part of this PR

Comment (blacha) [reply scripts/create_polygons.py]: this should be part of the PR, this PR changes the logic from exiting properly with a error code if something goes wrong to swallowing the errors.

**Review by MDavidson17 — APPROVED** [review]

**Review by amfage — APPROVED** [review]
